### DISCUSSION TITLE
Upload data async to HAPI

### DISCUSSION THREAD

**Discussion by ageryck** [opensrp/fhircore#3012] [Ideas]
Currently, the sync API operates on the principle of all or none this means if we have client-side resource changes that require sync up the API will attempt to batch all in a bundle that can entirely either succeed if all checks are passed or fail. This hinders data sync to the server side from this client completely until the issue is resolved from the client side, this issue is to try to figure out ways of changing the sync mechanism to eliminate stopping sync completely because of a few malformed/failing resources.

Options to discuss are;

1. Can we handle sync in batches and treat each batch as autonomous to the other such that if one fails it is marked as so and sync proceeds to the second batch
2. There's an option of syncing individual resources and treating individual fails separately from individual successes
3. Bundling related resources together for autonomous syncs

**Comment by pld** [comment]
(1) certainly makes sense, the other options sounds a bit more complex, (2) could make sense, but we'd expect lots of failure since bundles should reflect dependencies, but maybe those failures are OK, we don't lose anything with the failure (other than time), (3) would be a way to do (2) more effectively, if there's an existing libraries that creates a dependency graph of resources, and converts that graph into bundles, that'd be ideal.

@ellykits can you add some insight on how sync is done now, link to where that is in the code, and your thoughts on low effort ways to let sync complete for anything that does not fail (eg not fail all on single fail)

**Reply by ellykits** [reply]
FHIRCore schedules a background [AppSyncWorker](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/sync/AppSyncWorker.kt) job which is a subclass of the `FhirSyncWorker` provided via the SDK to run one time and periodic sync. The core functionalities for `download` and `upload` are internal to the  [FhirSyncWorker](https://github.com/google/android-fhir/blob/master/engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt) class.

The implementation for running sync is in the [SyncBroadcaster](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncBroadcaster.kt). The sync job statuses are then observed via the Activity/Fragments registered via the [SyncListenerManager](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/sync/SyncListenerManager.kt)


We do not override any functionality from the [FhirSynchronizer](https://github.com/google/android-fhir/blob/master/engine/src/main/java/com/google/android/fhir/sync/FhirSynchronizer.kt) class. The implementation for upload uses Bundle resource which is uploaded via a transaction. We confirmed the following from the SDK team:

1. The bundle is paginated by 500 Resources per request 
2. The upload runs in a transaction if one resource fails then the request fails
3. No dependency graph exists for related resources (SDK team commented that this should be possible to implement). Currently, the changes in the LocalEntityIndex table are all squashed per resource.

**Reply by ellykits** [reply]
I think our best chance would be to figure out how to sync dependent resources in batches. If a resource that other subsequent resources are dependent on fails to upload, the remaining requests will still fail.

**Reply by ellykits** [reply]
The SDK might need to expose some of the APIs to allow for modification/custom implementation.

**Reply by pld** [reply]
So I think a dependency graph is definitely the correct way to do this, but complex. If the goal is to get as many resource synced as possible, what if after a failed sync to do reference validation failure, we start syncing each resource independently, some will fail but some would succeeded, any failures during that sync we ignore.

I guess one big problem with this could be performance, but it seems worth trying, and from the user perspective better to sync some than none. I also wonder when we have sync blocked because of a ref validation failure, @ageryck around how many total resources remain unsynced? Oh, relatedly, when we have a ref validation sync failure, could we get HAPI to tells use specifically which resource failed validation? We could exclude, but of course that's not a great solution since other resources might be dependent on that resource and in which case there would still be a failure.

**Reply by ageryck** [reply]
@pld The dependency graph is an agreeable approach since it also eliminates incomplete data to HAPI as this may (not entirely) cause analytics challenges with any relational data. We spoke to Google in today's call and as Elly says each request queries for the top 500 records and pushes it as a bundle, if the bundle fails because of one resource the whole bundle fails and sync fails entirely since the next request will still query for the top 500 (the order observed is entry order to the table). For instance if we have 2300 records with an invalid record on row 1700 the the fourth bundle will fail and so up to 800 records will remain unsynced in the local db

**Reply by pld** [reply]
Yea dep graph is hard is the reason I'm looking for alternatives, buy maybe it's not that hard. Trying a resource at time is easy (idk, in theory in my head at least :)), a similar compromise would be continuing to try with the next bundle after failure, but the chances of having "aligned bundles" where all the resources in the bundle just happen to not conflict with the resources in the previously tried and failed bundle seems very low.

**Reply by ageryck** [reply]
Or if there can be a way to activate the dep graph only after the bundle fail, so in my example above sync batches of 500, and on the fourth bundle fails try organize it in dep graphs and in case of any success proceed to the next records in batches again to improve efficiency but maybe this is complicating it further

**Reply by pld** [reply]
@ageryck in practice, in the sync failures you've seen, about how many total resources remain unsynced on device?

**Reply by pld** [reply]
looking through the code, it is clear the change I suggested would be non-trivial, since we would have to implement our own sync worker, it looks like we could use the `UploadRequestGeneratorFactory` in `UrlRequest` to upload individual files, but still non-trivial. See if Google is interested in putting a depndency graph based sync approach on their roadmap. The plan I'd see for that, to consolidate would be:

1. create a dependency graph of all unsynced resources
2. iterate through the graph and for each tree create a bundle
3. sync each bundle independently in a transaction, using the existing transaction generation logic

**Reply by ageryck** [reply]
@pld This varies across devices we have devices that have over 800 tasks resources (this is an absolute count of create and update payloads), 250 careplans, 135 conditions etc so this can be as few as two tasks to as many as possible depending on when sync stopped working and how much more data has been collected afterward. Here a sample resources after sync failure 
<img src="https://github.com/opensrp/fhircore/assets/23252704/a8e11399-5c12-4637-80f4-165dede2d2dd" width="300" height="600">

**Reply by ellykits** [reply]
> See if Google is interested in putting a depndency graph based sync approach on their roadmap. The plan I'd see for that, to consolidate would be:
> 1. create a dependency graph of all unsynced resources
> 2. iterate through the graph and for each tree create a bundle
> 3. sync each bundle independently in a transaction, using the existing transaction generation logic

With the unsynced resource dependency graph available we can use the `UrlRequest` to send single requests for each bundle generated in step 2.

**Comment by ageryck** [comment]
Exploring options of uploading all data to HAPI then running validations server-side as documented [here](https://docs.google.com/document/d/1I1bBMwXXpzPBv5DxB-nRb968Qop3nOcs-kLbZrtvxJM/edit)
